include/graaflib/algorithm/strongly_connected_components/kosaraju.tpp

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <graaflib/algorithm/graph_traversal/depth_first_search.h>
+#include <graaflib/algorithm/strongly_connected_components/kosaraju.h>
+#include <graaflib/algorithm/utils.h>
+
+#include <iostream>
+#include <stack>
+
+namespace graaf::algorithm {
+
+template <typename V, typename E>
+std::vector<std::vector<vertex_id_t>> kosarajus_strongly_connected_components(
+    const directed_graph<V, E>& graph) {
+  std::vector<std::vector<vertex_id_t>> sccs;
+
+  if (graph.get_vertices().size() == 0) {
+    return sccs;
+  }
+
+  std::stack<vertex_id_t> stack;
+  std::unordered_set<vertex_id_t> seen_vertices;
+
+  std::function<void(vertex_id_t)> visit_vertex;
+  visit_vertex = [&](vertex_id_t vertex) {
+    if (!seen_vertices.contains(vertex)) {
+      stack.push(vertex);
+      seen_vertices.insert(vertex);
+
+      for (const auto neighbour : graph.get_neighbors(vertex)) {
+        visit_vertex(neighbour);
+      }
+    }
+  };
```

Comment:
Interesting, that sounds good! Could I ask why we would want to wrap it into another namespace here as opposed to just having it as a standalone function in the `.tpp` file - does this just signify that only functions in the `.tpp` file are able to access it and that users of the header won't be able to call the helper? 

Also, with situations like this would we still do a definition of the helper function in the header and have the definition in the `.tpp`?

Reply:
You are correct that adding it to the anonymous namespace will make it unavailable to any other compilation units. So we are not polluting the namespace of whoever includes this file.

In general I would restrict the scope of any helper functions as much as possible and only make it available on the interface if absolutely needed